modules/node-shared/src/main/scala/org/tessellation/node/shared/infrastructure/snapshot/GlobalSnapshotStateChannelEventsProcessor.scala

```diff
@@ -67,21 +66,65 @@ object GlobalSnapshotStateChannelEventsProcessor {
     new GlobalSnapshotStateChannelEventsProcessor[F] {
       private val logger = Slf4jLogger.getLoggerFromClass[F](GlobalSnapshotStateChannelEventsProcessor.getClass)
 
+      def deserialize[A: Decoder](binary: Signed[StateChannelSnapshotBinary]): F[Option[A]] =
+        jsonBrotliBinarySerializer.deserialize[A](binary.value.content).map(_.toOption)
+
+      def buildSnapshotFeesInfo(
+        lastGlobalSnapshotInfo: GlobalSnapshotInfo,
+        event: StateChannelOutput,
+        allFeesAddresses: Map[Address, Set[Address]]
+      ): F[SnapshotFeesInfo] =
+        event.snapshotBinary.value.lastSnapshotHash match {
+          case hash if hash == Hash.empty => SnapshotFeesInfo.empty.pure // genesis
+          case _ =>
+            deserialize[Signed[CurrencyIncrementalSnapshot]](event.snapshotBinary).flatMap {
+              case None => new Exception("Could not get snapshot after deserializing").raiseError[F, SnapshotFeesInfo]
+              case Some(snapshot) =>
+                Async[F].delay {
+                  val stakingBalance = fetchStakingBalance(event.address, lastGlobalSnapshotInfo)
+                  val maybeOwnerAddress = snapshot.value.messages.flatMap(_.find(_.messageType == MessageType.Owner)).map(_.address)
+                  val maybeStakingAddress = snapshot.value.messages.flatMap(_.find(_.messageType == MessageType.Staking)).map(_.address)
+                  SnapshotFeesInfo(allFeesAddresses, stakingBalance, maybeOwnerAddress, maybeStakingAddress)
+                }
+            }
+        }
+
       def process(
         snapshotOrdinal: SnapshotOrdinal,
         lastGlobalSnapshotInfo: GlobalSnapshotInfo,
         events: List[StateChannelOutput],
         validationType: StateChannelValidationType
-      )(implicit hasher: Hasher[F]): F[StateChannelAcceptanceResult] =
+      )(implicit hasher: Hasher[F]): F[StateChannelAcceptanceResult] = {
+        val allFeesAddresses: Map[Address, Set[Address]] =
+          lastGlobalSnapshotInfo.lastCurrencySnapshots.collect {
+            case (address, Right((_, info))) => Map(address -> info.lastMessages.toList.flatMap(_.values.map(_.address).toList).toSet)
+          }.flatten.toMap
```

Comment:
I have already seen that code 🤔 
Can we make it DRY somehow?

Maybe we can make `CurrencyMessageValidator` dependent on `LastSnapshotStorage`?
Or at least make a method somewhere like `getFeeAddresses(info: GlobalSnapshtoInfo)`

Reply:
Sure, I've update to have the function
`getFeeAddresses`